// node-htmldiff/htmldiff-cli.js
#!/usr/bin/env node
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fs_1 = require("fs");
var path_1 = require("path");
var diffHTML = require("./js/htmldiff");
// Argument count.
var argc = process.argv.length;
// Name of the "executable", e. g. `htmldiff-cli`.
var cli = path_1.parse(__filename).name;
// Optional class attribute name for `<ins>` and `<del>` tags.
var className = null;
// Optional prefix for data-operation-index attribute.
var dataPrefix = null;
// Optional list of atomic tags.
var atomicTags = null;
/**
 * Output usage info and version on the console.
 */
function printUsage() {
    var usage = cli + " v0.9.3\n\nUsage: " + cli + " beforeFile afterFile diffedFile [Options]\n\nbeforeFile:\n  An HTML input file in its original form.\n\nafterFile:\n  An HTML input file, based on beforeFile but with changes.\n\ndiffedFile:\n  Name of the diffed HTML output file. All differences between beforeFile\n  and afterFile will be surrounded with <ins> and <del> tags. If diffedFile\n  is - (minus) the result will be written with console.log() to stdout.\n\nOptions:\n\n-c className (optional):\n  className will be added as a class attribute on every <ins> and <del> tag.\n\n-p dataPrefix (optional):\n  The data prefix to use for data attributes. The operation index data\n  attribute will be named \"data-${dataPrefix-}operation-index\". If not\n  used, the default attribute name \"data-operation-index\" will be added\n  on every <ins> and <del> tag. The value of this attribute is an auto\n  incremented counter.\n\n-t atomicTags (optional):\n  Comma separated list of tag names. The list has to be in the form\n  \"tag1,tag2,...\" e. g. \"head,script,style\". An atomic tag is one whose\n  child nodes should not be compared - the entire tag should be treated\n  as one token. This is useful for tags where it does not make sense to\n  insert <ins> and <del> tags. If not used, this default list will be used:\n  \"iframe,object,math,svg,script,video,head,style\".";
    console.log(usage);
}
/**
 * Read content of a file and return content as a string. The encoding of the
 * file is considered to be in UTF-8 format by default.
 *
 * @param {string} fileName - The name of the file to be read.
 *
 * @returns {string} - The content of 'fileName' as a string. An empty string is
 *     considered to be an error which will be logged.
 */
function readFileContent(fileName) {
    try {
        var result = fs_1.readFileSync(path_1.resolve(fileName), "utf-8");
        if (result === "") {
            console.error("File \"" + fileName + "\" is empty.");
        }
        return result;
    }
    catch (error) {
        console.error("Couldn't read file \"" + fileName + "\"\n" + error.stack);
        return "";
    }
}
/**
 * Set the value of one the variables for the `diff` function based on a given
 * switch name. The function fails (return false) if the value of the variable
 * is already set (duplicate switch) or the switch name is unknown.
 *
 * @param name {string} - The name of the switch (one of `-c`, `-p` or `-t`)
 * @param value {string} - The value of the switch given on the command line.
 *
 * @returns {boolean} - `true` if the switch could be resolved, otherwise `false`.
 */
function resolveSwitch(name, value) {
    switch (name) {
        case "-c":
            if (className) {
                return false;
            }
            className = value;
            return true;
        case "-p":
            if (dataPrefix) {
                return false;
            }
            dataPrefix = value;
            return true;
        case "-t":
            if (atomicTags) {
                return false;
            }
            atomicTags = value;
            return true;
    }
    // Unknown switch
    return false;
}
// Invalid argc
var validArgc = [5, 7, 9, 11];
if (validArgc.indexOf(argc) === -1) {
    printUsage();
    process.exit(1);
}
// Resolve switches
if (argc > 6) {
    if (!resolveSwitch(process.argv[5], process.argv[6])) {
        printUsage();
        process.exit(1);
    }
}
if (argc > 8) {
    if (!resolveSwitch(process.argv[7], process.argv[8])) {
        printUsage();
        process.exit(1);
    }
}
if (argc > 10) {
    if (!resolveSwitch(process.argv[9], process.argv[10])) {
        printUsage();
        process.exit(1);
    }
}
// Execute diff
var beforeFile = readFileContent(process.argv[2]);
if (!beforeFile) {
    process.exit(1);
}
var afterFile = readFileContent(process.argv[3]);
if (!afterFile) {
    process.exit(1);
}
var diffedResult = diffHTML(beforeFile, afterFile, className, dataPrefix, atomicTags);
if (process.argv[4] === "-") {
    console.log(diffedResult);
}
else {
    fs_1.writeFileSync(path_1.resolve(process.argv[4]), diffedResult);
}
